src/optimizer/index.js

```diff
@@ -30,13 +32,14 @@ module.exports = {
       ? transformsWhitelist
       : Object.keys(optimizationTransforms);
 
+    let result = new transform.TransformResult(parser.parse(regexp));
     let prevResult;
-    let result;
+    let ast;
+
     do {
-      if (result) {
-        prevResult = result.toString();
-        regexp = prevResult;
-      }
+      prevResult = result.toString();
+      ast = clone(result.getAST());
```

Comment:
We'll need to check for speed regressions. When the optimizer is used on a huge cobe base, this might be a potential performance regression. Do you know how fast/slow this `clone` is?

Reply:
I don't, no. 
The thing is I can't rollback a transform if I can't restore the AST to its previous state. And since `transform()` mutates the AST, I do need a copy here.

I might try to make some kind of benchmark with different versions of regexp-tree.